parser/openfoodfacts_taxonomy_parser/parser/parser.py

```diff
@@ -1,158 +1,213 @@
+        )
+
+    def _create_other_nodes(self, other_nodes: list[NodeData], project_label: str):
+        """Create all TEXT, SYNONYMS and STOPWORDS nodes"""
+        self.parser_logger.info("Creating TEXT, SYNONYMS and STOPWORDS nodes")
+        start_time = timeit.default_timer()
+
+        with self.session.begin_transaction() as tx:
+            for node in other_nodes:
+                self._create_other_node(tx, node, project_label)
 
-    def create_previous_link(self, previous_links: list[PreviousLink], multi_label: str):
+        self.parser_logger.info(
+            f"Created {len(other_nodes)} TEXT, SYNONYMS and STOPWORDS nodes in {timeit.default_timer() - start_time} seconds"
+        )
+
+    def _create_previous_links(self, previous_links: list[PreviousLink], project_label: str):
+        """Create the 'is_before' relations between nodes"""
         self.parser_logger.info("Creating 'is_before' links")
-        for previous_link in previous_links:
-            id = previous_link["id"]
-            before_id = previous_link["before_id"]
-
-            query = f"""
-                MATCH(n:{multi_label}) WHERE n.id = $id
-                MATCH(p:{multi_label}) WHERE p.id= $before_id
-                CREATE (p)-[r:is_before]->(n)
-                RETURN r
-            """
-            results = self.session.run(query, id=id, before_id=before_id)
-            relation = results.values()
-            if len(relation) > 1:
-                self.parser_logger.error(
-                    "2 or more 'is_before' links created for ids %s and %s, "
-                    "one of the ids isn't unique",
-                    id,
-                    before_id,
-                )
-            elif not relation[0]:
-                self.parser_logger.error("link not created between %s and %s", id, before_id)
-
-    def create_child_link(self, child_links: list[ChildLink], multi_label: str):
-        """Create the relations between nodes"""
+        start_time = timeit.default_timer()
+
+        query = f"""
+            UNWIND $previous_links as previous_link
+            MATCH(n:{project_label}) USING INDEX n:{project_label}(id)
+            WHERE n.id = previous_link.id
+            MATCH(p:{project_label}) USING INDEX p:{project_label}(id)
+            WHERE p.id = previous_link.before_id
+            CREATE (p)-[relations:is_before]->(n)
+            WITH relations
+            UNWIND relations AS relation
+            RETURN COUNT(relation)
+        """
+        result = self.session.run(query, previous_links=previous_links)
+        count = result.value()[0]
+
+        self.parser_logger.info(
+            f"Created {count} 'is_before' links in {timeit.default_timer() - start_time} seconds"
+        )
+
+        if count != len(previous_links):
+            self.parser_logger.error(
+                f"Created {count} 'is_before' links, {len(previous_links)} links expected"
+            )
```

Comment:
I'm asking myself if it's not a show stopper. This is a very important part to be able to unparse the taxonomy. So it would be a pity we let a user edit a taxonomy that won't be rendered as a PR in the end !

We should at least add it to the error nodes of the taxonomy.

Reply:
This is already added to the error node of the taxonomy.
I agree : errors in general should be fixed before modifying the taxonomy. The frontend should better reflect that (https://github.com/openfoodfacts/taxonomy-editor/issues/332)